src/server/server_request_responder.hpp

```diff
@@ -120,17 +179,70 @@ void
 server_request_responder<T_CONN>::async_execute (connection_t &a_conn, sequenced_payload_t &&a_sp,
     handler_func_t &&a_func)
 {
-  async_execute (new task (a_conn, std::move (a_sp), std::move (a_func)));
+  new_task (&a_conn);
+  async_execute (new task (*this, a_conn, std::move (a_sp), std::move (a_func)));
+}
+
+template<typename T_CONN>
+void
+server_request_responder<T_CONN>::new_task (const connection_t *connection_ptr)
+{
+  std::lock_guard<std::mutex> lockg { m_executing_tasks_mtx };
+  connection_executing_task_type &executing_task_for_connection = m_executing_tasks[connection_ptr];
+  ++executing_task_for_connection.n_count;
+}
+
+template<typename T_CONN>
+void
+server_request_responder<T_CONN>::retire_task (const connection_t *connection_ptr)
+{
+  {
+    std::lock_guard<std::mutex> lockg { m_executing_tasks_mtx };
+    connection_executing_task_type &executing_task_for_connection = m_executing_tasks[connection_ptr];
+    --executing_task_for_connection.n_count;
+    assert (executing_task_for_connection.n_count >= 0);
+    if (executing_task_for_connection.n_count == 0)
+      {
+	// notify from under the lock to avoid possibility of a race condition with the
+	// connection being waited to become idle (see corresponding function) which also - if
+	// successful - removes the entry from the container
+	executing_task_for_connection.m_cv.notify_one ();
+      }
+  }
+}
+
+template<typename T_CONN>
+void
+server_request_responder<T_CONN>::wait_connection_to_become_idle (const connection_t *connection_ptr)
+{
+  constexpr std::chrono::milliseconds millis_20 { 20 };
+
+  std::unique_lock<std::mutex> ulock { m_executing_tasks_mtx };
+  auto found_it = m_executing_tasks.find (connection_ptr);
+  if (found_it != m_executing_tasks.end ())
+    {
+      connection_executing_task_type &executing_task_for_connection = (*found_it).second;
+      while (true)
+	{
+	  if (executing_task_for_connection.m_cv.wait_for (ulock, millis_20,
+	      [&executing_task_for_connection] { return (executing_task_for_connection.n_count == 0); }))
+	    {
+	      m_executing_tasks.erase (found_it);
```

Comment:
Incorrect comment because the responder is only waited for once to become idle for every connection. Refactored the code to formally enforce this invariant.